network/lwip2transport/device.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lwip2transport
+
+import (
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-internal-sdk/network"
+	"github.com/Jigsaw-Code/outline-internal-sdk/transport"
+	lwip "github.com/eycorsican/go-tun2socks/core"
+)
+
+const packetMTU = 1500
+
+// LwIPTransportDevice is a [network.IPDevice] that can translate IP packets to TCP/UDP traffic and vice versa. It uses
+// the [lwIP library] to perform the translation.
+//
+// LwIPTransportDevice must be a singleton object due to limitations in [lwIP library].
+//
+// To use a LwIPTransportDevice:
+//  1. Call [ConfigureDevice] with two handlers for TCP and UDP traffic.
+//  2. Write IP packets to the device. The device will translate the IP packets to TCP/UDP traffic and send them to the
+//     appropriate handlers.
+//  3. Read IP packets from the device to get the TCP/UDP responses.
+//
+// A LwIPTransportDevice is NOT thread-safe. However it is safe to use Write, Read/WriteTo and Close in different
+// goroutines. But keep in mind that only one goroutine can call Write at a time; and only one goroutine can use either
+// Read or WriteTo at a time.
+//
+// [lwIP library]: https://savannah.nongnu.org/projects/lwip/
+type LwIPTransportDevice interface {
+	network.IPDevice
+}
+
+// Compilation guard against interface implementation
+var _ network.IPDevice = (*lwIPDevice)(nil)
+var _ LwIPTransportDevice = (*lwIPDevice)(nil)
+
+type lwIPDevice struct {
+	tcp   *tcpHandler
+	udp   *udpHandler
+	stack lwip.LWIPStack
+
+	// whether the device has been closed
+	done chan struct{}
+
+	// async read call and its result
+	rdBuf chan []byte
+	rdN   chan int
+}
+
+// Singleton instance
+var instMu sync.Mutex
+var inst *lwIPDevice = nil
+
+// ConfigureDevice configures the singleton LwIPTransportDevice using the [transport.StreamDialer] sd to handle TCP
+// streams and the [transport.PacketListener] pl to handle UDP packets.
+//
+// You can only have one active LwIPTransportDevice per process. If you try to call ConfigureDevice more than once, we
+// will Close the previous device and reconfigures it.
+func ConfigureDevice(sd transport.StreamDialer, pl transport.PacketListener) (LwIPTransportDevice, error) {
```

Comment:
LwIPTransportDevice doesn't add any methods, so you might as well return an IPDevice and remove that type.

You can add the type comments here.

Reply:
Right, I was planning for the future. But we can remove it in this PR.